### ISSUE TITLE
Query: Apply CompositePredicateExpressionVisitor to join predicates

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#9395]
At present it is applied to `SelectExpression.Predicate` only

**Comment by maumar** [comment]
Also add more optimizations, e.g. (`(c.Discriminator == "Foo" ? c.FK : null) == o.PK` into `c.Discriminator == "Foo" && c.FK == o.PK`)

**Comment by ajcvickers** [comment]
@maumar @smitpatel Please update description with more details and investigate for 2.1.

**Comment by smitpatel** [comment]
Derived Include generates join condition using ternary to appropriately cast to derived type. 
Which generates query like this
```SQL
SELECT [l].[Name], [l].[Discriminator], [l].[LocustHordeId], [l].[ThreatLevel], [l].[DefeatedByNickname], [l].[DefeatedBySquadId], [t].[Nickname], [t].[SquadId], [t].[AssignedCityName], [t].[CityOrBirthName], [t].[Discriminator], [t].[FullName], [t].[HasSoulPatch], [t].[LeaderNickname], [t].[LeaderSquadId], [t].[Rank]
FROM [LocustLeaders] AS [l]
LEFT JOIN (
    SELECT [l.DefeatedBy].*
    FROM [Gears] AS [l.DefeatedBy]
    WHERE [l.DefeatedBy].[Discriminator] IN (N'Officer', N'Gear')
) AS [t] ON (CASE
    WHEN [l].[Discriminator] = N'LocustCommander'
    THEN [l].[DefeatedByNickname] ELSE NULL
END = [t].[Nickname]) AND (CASE
    WHEN [l].[Discriminator] = N'LocustCommander'
    THEN [l].[DefeatedBySquadId] ELSE NULL
END = [t].[SquadId])
WHERE [l].[Discriminator] IN (N'LocustCommander', N'LocustLeader')
```
Here using Case block on join condition means join cannot use any index and no longer sargable.

**Comment by maumar** [comment]
fixed in 9355f64dbc5846271c36bbf6ff8196ed954751fe
